### PR TITLE
feat(validator-node): initial state sync implementation (partial)

### PR DESCRIPTION
**Pull request by sdbondi**
Description
---
- add OpLogs
- add comms RPC methods and client calls to sync validator node state
- add synchronising state and basic implementation to validator node state machine (currently always in sync due to incorrect Merkle root in initial checkpoint)
- fix: fetching checkpoint from base node returned incorrect field for asset public key
- add read only interface for state db

Motivation and Context
---
Validator node on a committee needs to sync their state after going offline.

How Has This Been Tested?
---
Manually, it didnt work because the merkle root of the initial checkpoint is incorrect. Sync component has been hardcoded to always say it is in sync for now.

### CODE REVIEW

File: applications/tari_validator_node/proto/dan/validator_node.proto [hunk 1]
```diff
@@ -74,7 +74,48 @@ message GetSidechainBlocksRequest {
...
+  uint64 height = 2;
+}
+
+message GetStateOpLogsResponse {
+  repeated StateOpLog op_logs = 1;
+}
+
+message StateOpLog {
+  uint64 height = 1;
+  string operation = 2;
```
Comment (stringhandler) [inline applications/tari_validator_node/proto/dan/validator_node.proto:108]: perhaps an enum here rather?

File: dan_layer/core/src/services/asset_processor.rs [hunk 2]
```diff
@@ -89,12 +89,12 @@ impl AssetProcessor for ConcreteAssetProcessor {
         )
     }
 
-    fn invoke_read_method<TUnifOfWork: StateDbUnitOfWork>(
+    fn invoke_read_method<TUnitOfWork: StateDbUnitOfWorkReader>(
```
Comment (stringhandler) [inline dan_layer/core/src/services/asset_processor.rs:92]: Nice. Perhaps we need a better name, but I like it

**Review by stringhandler — DISMISSED** [review]
Very nice

File: dan_layer/core/src/services/asset_processor.rs (hunk 2 above)
Comment (sdbondi) [reply dan_layer/core/src/services/asset_processor.rs:92]: Yeah it's a bit long, open to suggestions/change later

File: applications/tari_validator_node/proto/dan/validator_node.proto (hunk 1 above)
Comment (sdbondi) [reply applications/tari_validator_node/proto/dan/validator_node.proto:108]: Yeah probably, I'm using this "weirdly" as a single char enum - `S = Set, D = Delete` etc. but an enum is more standard :P

**Review by stringhandler — APPROVED** [review]

**Pull request merged**